### PR TITLE
core: do not create symlink to private directory if parent already exists

### LINKED ISSUE #10337: Ignore remote filesystems in test-execute
**systemd version the issue has been seen with**
> systemd-239

**Used distribution**
 > RHEL

**Expected behaviour you didn't see**
 > **test-execute** finishes without timeout

**Unexpected behaviour you saw**
 > **test-execute** asserts on a timeout

**Steps to reproduce the problem**

In our test infrastructure (at Red Hat) we have several large remote filesystems mounted by default under /mnt. The **test-execute** test case runs the [exec-dynamicuser-statedir.service](https://github.com/systemd/systemd/blob/master/test/test-execute/exec-dynamicuser-statedir.service#L16) which executes `find` command over the entire filesystem, including /mnt, which - in this case - results in a test timeout:

```
# meson test -v test-execute
...
++ find / '(' -path /var/tmp -o -path /tmp -o -path /proc -o -path /dev/mqueue -o -path /dev/shm -o -path /sys/fs/bpf ')' -prune -o -type d -writable -print
++ sort -u
++ tr -d '\n'
...
Test timeout when testing exec-dynamicuser-statedir.service
1/1 test-execute                            FAIL    122.43 s
...
# ps aux | grep find
...
65388    23592  1.9  0.1  21616  4344 ?        D    11:19   0:00 find / ( -path /var/tmp -o -path /tmp -o -path /proc -o -path /dev/mqueue -o -path /dev/shm -o -path /sys/fs/bpf ) -prune -o -type d -writable -print
...
```

I managed to temporarily workaround this issue by adding `-fstype nfs -o -fstype nfs4` to the `-prune` part of the condition[0], but that doesn't cover all possible remote filesystems. Another possibility would be to ignore `/mnt` altogether.

[0]
```diff
--- ../test/test-execute/exec-dynamicuser-statedir.service	2018-10-09 12:07:04.793759716 -0400
+++ exec-dynamicuser-statedir.service	2018-10-09 12:07:01.321759716 -0400
@@ -12,7 +12,7 @@
 ExecStart=test -f /var/lib/private/quux/pief/yayyay
 
 # Make sure that /var/lib/private/waldo is really the only writable directory besides the obvious candidates
-ExecStart=sh -x -c 'test $$(find / \( -path /var/tmp -o -path /tmp -o -path /proc -o -path /dev/mqueue -o -path /dev/shm -o -path /sys/fs/bpf \) -prune -o -type d -writable -print 2>/dev/null | sort -u | tr -d '\\\\n') = /var/lib/private/quux/pief/var/lib/private/waldo'
+ExecStart=sh -x -c 'test $$(find / \( -fstype nfs -o -fstype nfs4 -o -path /var/tmp -o -path /tmp -o -path /proc -o -path /dev/mqueue -o -path /dev/shm -o -path /sys/fs/bpf \) -prune -o -type d -writable -print 2>/dev/null | sort -u | tr -d '\\\\n') = /var/lib/private/quux/pief/var/lib/private/waldo'
 
 Type=oneshot
 DynamicUser=yes
```
Is any of the aforementioned solutions reasonable?

### LINKED ISSUE #24783: DynamicUser with no static User cannot create multiple StateDirectories that share a common prefix
### systemd version the issue has been seen with

251

### Used distribution

NixOS unstable

### Linux kernel version used

5.15.63

### CPU architectures issue was seen on

x86_64

### Component

systemctl, systemd

### Expected behaviour you didn't see

It creates multiple state directories, even if they do share a common prefix, by, e.g. sorting them or ignoring already existing prefixes.

### Unexpected behaviour you saw

The service fails with `status=238/STATE_DIRECTORY`.

### Steps to reproduce the problem

(1) Create a systemd service with `DynamicUser=true` and `User=user` where `user` is not static in the system (maybe static setups fails also, I did not test.)
(2) Add `StateDirectory=data data/uploads`, possible by manual pages and possible when `DynamicUser=false` for example

Run this service.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
machine # [    8.743574] systemd[945]: listmonk.service: Failed to set up special execution directory in /var/lib: File exists
machine # [    8.745210] nscd[880]: 880 monitoring directory `/etc` (2)
machine # [    8.745777] systemd[945]: listmonk.service: Failed at step STATE_DIRECTORY spawning /nix/store/bii7l73s4mldagsfrn86161xzsns1810-listmonk-2.2.0/bin/listmonk: File exists
machine # [    8.747025] nscd[880]: 880 monitoring file `/etc/passwd` (1)
machine # [    8.747586] systemd[1]: listmonk.service: Control process exited, code=exited, status=238/STATE_DIRECTORY
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #24783.
Closes #10337.

### CODE REVIEW

**Comment by yuwata** [comment]
Updated, tests are adjusted. PTAL.

**Comment by yuwata** [comment]
Hm, still red.

@mrc0mmand What does the following log mean? What should I do? I tried to add `UnsetEnvironment=LD_PRELOAD`, but it does not work.
```
[   20.905297] testsuite-34.sh[854]: + systemctl start testsuite-34-check-writable.service
         Starting [0;1;39mtestsuite-34-check-writable.service[0m...
[   20.987808] bash[1323]: ==1323==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
[[0;1;31mFAILED[0m] Failed to start [0;1;39mtestsuite-34-check-writable.service[0m.
See 'systemctl status testsuite-34-check-writable.service' for details.
```

**Comment by mrc0mmand** [comment]
> Hm, still red.
> 
> @mrc0mmand What does the following log mean? What should I do? I tried to add `UnsetEnvironment=LD_PRELOAD`, but it does not work.
> 
> ```
> [   20.905297] testsuite-34.sh[854]: + systemctl start testsuite-34-check-writable.service
>          Starting �[0;1;39mtestsuite-34-check-writable.service�[0m...
> [   20.987808] bash[1323]: ==1323==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
> [�[0;1;31mFAILED�[0m] Failed to start �[0;1;39mtestsuite-34-check-writable.service�[0m.
> See 'systemctl status testsuite-34-check-writable.service' for details.
> ```

Ah, this is unfortunate - we don't set `$LD_PRELOAD` globally anymore, because it caused a lot of issues in other places, so now only specific binaries are wrapped with shell script that does the magic. In this case, `DynamicUser=yes` pulls in systemd's nss libraries into the bash invocation, which then complains, because you need `$LD_PRELOAD` for unsanitized binaries if they pull in sanitized libs.

After adding `Envrironment=/lib64/libasan.so.8` to the `testsuite-34-check-writable.service` it seems to be working fine. The question is how to do that without hardcoding the DSO name. The environment already provides `$ASAN_RT_PATH` with the current ASan DSO, but you can't do `Environment=LD_PRELOAD=$ASAN_RT_PATH` to set it. 

Maybe creating an env file in the `create_asan_wrapper()` function (like `/usr/lib/systemd/systemd-asan-env` and passing it to services which need it via `EnvironmentFile=-/usr/lib/systemd/systemd-asan-env` might work - not sure if there's a nicer way.

**Comment by mrc0mmand** [comment]
Something like this seems to be working fine with both sanitized and regular builds:

```diff
diff --git a/test/test-functions b/test/test-functions
index fa3ad9fbdc..030e5ea514 100644
--- a/test/test-functions
+++ b/test/test-functions
@@ -849,7 +849,15 @@ fi
 
 DEFAULT_ASAN_OPTIONS=${ASAN_OPTIONS:-strict_string_checks=1:detect_stack_use_after_return=1:check_initialization_order=1:strict_init_order=1}
 DEFAULT_UBSAN_OPTIONS=${UBSAN_OPTIONS:-print_stacktrace=1:print_summary=1:halt_on_error=1}
-DEFAULT_ENVIRONMENT="ASAN_OPTIONS=\$DEFAULT_ASAN_OPTIONS UBSAN_OPTIONS=\$DEFAULT_UBSAN_OPTIONS LSAN_OPTIONS=\$DEFAULT_LSAN_OPTIONS"
+DEFAULT_ENVIRONMENT="ASAN_OPTIONS=\$DEFAULT_ASAN_OPTIONS UBSAN_OPTIONS=\$DEFAULT_UBSAN_OPTIONS"
+
+# Create a simple environment file which can be included by systemd services
+# that need it (i.e. service that utilize DynamicUser=true and bash, etc.)
+cat >/usr/lib/systemd/systemd-asan-env <<INNER_EOF
+LD_PRELOAD=$ASAN_RT_PATH
+ASAN_OPTIONS=$DEFAULT_ASAN_OPTIONS
+UBSAN_OPTIONS=$DEFAULT_UBSAN_OPTIONS
+INNER_EOF
 
 # As right now bash is the PID 1, we can't expect PATH to have a sane value.
 # Let's make one to prevent unexpected "<bin> not found" issues in the future
diff --git a/test/units/testsuite-34-check-writable.service b/test/units/testsuite-34-check-writable.service
index a84701e558..28c3975552 100644
--- a/test/units/testsuite-34-check-writable.service
+++ b/test/units/testsuite-34-check-writable.service
@@ -3,7 +3,7 @@ Description=Check writable directories when DynamicUser= with StateDirectory=
 
 [Service]
 # Relevant only for sanitizer runs
-UnsetEnvironment=LD_PRELOAD
+EnvironmentFile=-/usr/lib/systemd/systemd-asan-env
 
 Type=oneshot
 DynamicUser=yes
```

**Comment by yuwata** [comment]
@mrc0mmand Thank you for your help! The suggested changes are applied.

**Comment by yuwata** [comment]
Hm,,, ubuntu CIs are unhappy...

**Comment by mrc0mmand** [comment]
what the hell is going on with kernel recently:

```

File: src/core/execute.c [hunk 1]
```diff
@@ -7040,33 +7040,56 @@ void exec_directory_done(ExecDirectory *d) {
...
-        assert(n);
         assert(path);
 
+        existing = exec_directory_find(d, path);
+        if (existing) {
+                if (symlink) {
+                        r = strv_extend(&existing->symlinks, symlink);
+                        if (r < 0)
+                                return r;
+                }
```
Comment (poettering) [inline src/core/execute.c]: you can drop this conditionalization. strv_extend() is fine with the second param being NULL, and then is a NOP

File: src/core/execute.c [hunk 2]
```diff
@@ -2412,12 +2413,14 @@ static int setup_exec_directory(
...
-                        if (r < 0)
-                                goto fail;
+                        if (!context->directories[type].items[i].only_create) {
+                                /* And link it up from the original place. Note that if a mount namespace is going to be
+                                 * used, then this symlink remains on the host, and a new one for the child namespace will
+                                 * be created later. */
+                                r = symlink_idempotent(pp, p, true);
+                                if (r < 0)
+                                        goto fail;
+                        }
```
Comment (poettering) [inline src/core/execute.c:2433]: I don't grok this change. `only_create` means we only do mkdir(), but not symlink(), did I get that right?

But what does that have to do with the issue this is supposed to fix? i can't see the relationship? that issue doesn't even use symlinks? what am I missing? can you explain?

File: src/core/execute.c [hunk 3]
```diff
@@ -7092,6 +7101,29 @@ int exec_directory_add(ExecDirectory *d, const char *path, const char *symlink)
...
+        assert(d);
+
+        if (d->n_items <= 1)
+                return;
+
+        typesafe_qsort(d->items, d->n_items, exec_directory_item_compare_func);
+
+        for (size_t i = 1; i < d->n_items; i++)
+                for (size_t j = 0; j < i; j++)
+                        if (path_startswith(d->items[i].path, d->items[j].path)) {
```
Comment (poettering) [inline src/core/execute.c:7136]: hmm, so you are checking any of the earlier items are a prefix of the current one, iiuc? but why does that matter for symlink creation? not following...

File: src/core/execute.h [hunk 4]
```diff
@@ -135,6 +135,7 @@ typedef enum ExecDirectoryType {
 typedef struct ExecDirectoryItem {
         char *path;
         char **symlinks;
+        bool only_create;
 } ExecDirectoryItem;
```
Comment (poettering) [inline src/core/execute.h:139]: so far we tried to keep runtime data and configured settings in seperate structs (though there are exceptions, i.e. ExecCommand is config data but embedds ExecStatus which is runtime data). Which makes me a bit concerned about embedding this bool here given it's only used internally, for a brief moment even...

**Comment by poettering** [comment]
hmm, the explanation i asked for should probably end up in comments in the code, and not here, actually ;-)

File: src/core/execute.c (hunk 2 above)
Comment (yuwata) [reply src/core/execute.c:2433]: E.g. `StateDirectory=foo foo/bar`, then when processing `"foo/bar"`, `pp = "/var/lib/private/foo/bar"` and `p = "/var/lib/foo/bar"`. But there already exists a symlink `/var/lib/private/foo -> /var/lib/foo`. So, the inode of `pp` and `p` are the same, and we cannot create the symlink.

I added comments in the code.

File: src/core/execute.c (hunk 3 above)
Comment (yuwata) [reply src/core/execute.c:7136]: See the above comment.

File: src/core/execute.h (hunk 4 above)
Comment (yuwata) [reply src/core/execute.h:139]: I do not think this is runtime data. This is a kind of preprocessed data from unit files and/or dbus operation for code optimization, and purely calculated from them. This just caches the existence of one of parent directories in the list.
Without the flag, we need to search the list twice in `setup_exec_directory()` and `compile_bind_mounts()`.

**Comment by yuwata** [comment]
@poettering Comments with example are added, PTAL.

**Comment by yuwata** [comment]
@bluca and @poettering Could you review this again?

**Comment by keszybz** [comment]
LGTM.

**Pull request merged**